### PR TITLE
[Refactor] Refactoring of the saving process in exercise-update-component

### PR DESCRIPTION
**Pull request by Olimbia23**
### Checklist
- [X] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [X] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).

### Motivation and Context
After working with `exercise-update-component.ts` files, I have noticed some duplication. One of them revolves around save method. This PR moves the save logic to a separate class and calls it from these components, removing duplication.

### Description

I introduce a new class called `SaveExerciseCommand<T extends Exercise>` which performs the save for the specified `Exercise` as well as a `ExerciseServicable<T extends Exercise>` that is an interface that other ExerciseServices conform to.

### Steps for testing

1. Log in to Artemis
2. Navigate to Course Management
3. Pick a course
4. Create new text-/modeling-/file-upload exercises
5. Creating/Updating the exercises should work as in develop

### Test Coverage
Exercise-Utils: 94.95%

### CODE REVIEW

**Review by argertboja — COMMENTED** [review]
I faced no issues while testing on TS1 👍🏼

**Review by maximilianbergmann — DISMISSED** [review]
Tested on TS3. Works normal

**Review by real-dmy — APPROVED** [review]
Tested on TS1. No issues.

**Review by joschahenningsen — APPROVED** [review]
Changes look good, had no issues either :)

**Review by coolchock — APPROVED** [review]
tested on ts3, works as expected :+1:

**Review by jhainzlschmid — DISMISSED** [review]
Tested on ts1

**Review by jpbernius — DISMISSED** [review]

**Review by sinangunduz — APPROVED** [review]
tested on ts1, had no issues

**Comment by krusche** [comment]
@Olimbia23 there is a merge conflict. Could you please resolve it and ask for re-approvals?

**Review by ArtemTsurikov — DISMISSED** [review]
Tested on TS1, looks good!

**Review by canyelten — APPROVED** [review]
Tested on TS1, works as expected.

**Review by katjanakosic — DISMISSED** [review]
Everything worked fine on ts2

File: src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts [hunk 1]
```diff
@@ -47,6 +49,10 @@ export class FileUploadExerciseUpdateComponent implements OnInit {
         private router: Router,
     ) {}
 
+    get editType(): EditType {
+        return this.fileUploadExercise.id === undefined ? EditType.CREATE : EditType.UPDATE;
```
Comment (FrankeLukas) [inline src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts:53]:
```suggestion
        return this.fileUploadExercise.id == undefined ? EditType.CREATE : EditType.UPDATE;
```

File: src/test/javascript/spec/component/modeling-exercise/modeling-exercise-update.component.spec.ts [hunk 2]
```diff
@@ -266,18 +275,4 @@ describe('ModelingExercise Management Update Component', () => {
...
-        comp.modelingExercise = new ModelingExercise(UMLDiagramType.Flowchart, undefined, undefined);
-        const exampleSubmission1 = new ExampleSubmission();
-        exampleSubmission1.id = 1;
-        const exampleSubmission2 = new ExampleSubmission();
-        exampleSubmission2.id = 2;
-        comp.modelingExercise.exampleSubmissions = [exampleSubmission1, exampleSubmission2];
-        spyOn(exampleSubmissionService, 'delete').and.returnValue(of(new HttpResponse({})));
-        comp.deleteExampleSubmission(2, 1);
-        tick();
-        expect(exampleSubmissionService.delete).toHaveBeenCalledWith(2);
```
Comment (FrankeLukas) [inline src/test/javascript/spec/component/modeling-exercise/modeling-exercise-update.component.spec.ts:280]: This deletes the last usage of `exampleSubmissionService`, so the variable (and import) itself should also be removed I suppose?

**Review by ivanchimeno — DISMISSED** [review]

**Review by NicolasUllmann — DISMISSED** [review]
Tested on TS2. Everything worked as expected.

**Review by imstreetspirit — CHANGES_REQUESTED** [review]
tested on ts2. I created new modeling/text/file exercises and updated. it works. But I also tested for re-evaluation. So I submitted student solutions and assessed them. During the assessment I used grading instructions for feedback. After that I go back the exercise edit view, I change credit/maxCountUsage and delete one of them grading instruction in there, then I clicked save button. In this case, there should be exercise-update-warning popup, but it did not open the popup. Instead of, it automatically updated the exercise with grading instructions. This happens for each type of exercise. I can help if this explanation is not enough to describe the problem

File: src/main/webapp/app/exercises/shared/exercise/exercise-utils.ts [hunk 3]
```diff
@@ -7,13 +7,79 @@ import { ProgrammingExercise } from 'app/entities/programming-exercise.model';
...
+                        return this.exerciseService.reevaluateAndUpdate(ex, requestOptions);
+                    } else {
+                        return this.exerciseService.update(ex, requestOptions);
+                    }
+            }
+        };
+
+        let saveObservable = of([false, prepareRequestOptions()]);
+
+        if (exercise.gradingInstructionFeedbackUsed && this.modalService.hasOpenModals()) {
```
Comment (imstreetspirit) [inline src/main/webapp/app/exercises/shared/exercise/exercise-utils.ts:60]: maybe the problem is related with this check. In the current version, first I check `exercise.gradingInstructionFeedbackUsed` is true or not after that, invoke `this.popupService.checkExerciseBeforeUpdate`... this one decides whether the popup can be opened or not.. So, `this.modalService.hasOpenModals()` check should come after that... I will try on my local, and let you know...

File: src/main/webapp/app/exercises/shared/exercise/exercise-utils.ts [hunk 4]
```diff
@@ -7,13 +7,79 @@ import { ProgrammingExercise } from 'app/entities/programming-exercise.model';
...
+            );
+            const reEvaluatedCase = popupRefObs.pipe(
+                mergeMap((ref) =>
+                    (ref.componentInstance as ExerciseUpdateWarningComponent).reEvaluated.pipe(map(() => [true, { deleteFeedback: ref.componentInstance.deleteFeedback }])),
+                ),
+            );
+            const canceledCase = popupRefObs.pipe(mergeMap((ref) => (ref.componentInstance as ExerciseUpdateWarningComponent).canceled));
+
+            saveObservable = confirmedCase.pipe(mergeWith(reEvaluatedCase), takeUntil(canceledCase));
+        }
```
Comment (imstreetspirit) [inline src/main/webapp/app/exercises/shared/exercise/exercise-utils.ts:74]:
```suggestion
        if (exercise.gradingInstructionFeedbackUsed) {
            const popupRefObs = from(this.popupService.checkExerciseBeforeUpdate(exercise, this.backupExercise));
            
            if (this.modalService.hasOpenModals()) {
                const confirmedCase = popupRefObs.pipe(
                    mergeMap((ref) => (ref.componentInstance as ExerciseUpdateWarningComponent).confirmed.pipe(map(() => [false, prepareRequestOptions()]))),
                );
                const reEvaluatedCase = popupRefObs.pipe(
                    mergeMap((ref) =>
                        (ref.componentInstance as ExerciseUpdateWarningComponent).reEvaluated.pipe(map(() => [true, { deleteFeedback: ref.componentInstance.deleteFeedback }])),
                    ),
                );
                const canceledCase = popupRefObs.pipe(mergeMap((ref) => (ref.componentInstance as ExerciseUpdateWarningComponent).canceled));

                saveObservable = confirmedCase.pipe(mergeWith(reEvaluatedCase), takeUntil(canceledCase));
            }
        }
```

Comment (imstreetspirit) [reply src/main/webapp/app/exercises/shared/exercise/exercise-utils.ts:74]: tried it locally, it works

**Review by cremertim — DISMISSED** [review]
worked on ts 1 as expected :) //despite the error mentioned by Merve

**Comment by Olimbia23** [comment]
Thank you @imstreetspirit. Now should work as expected.

**Review by imstreetspirit — APPROVED** [review]
tested on ts2 again. great job!! thanks for quick fix! reevaluation works properly again.

**Review by willergeorg — APPROVED** [review]
Tested on TS2. Worked good for me.

**Review by NicolasUllmann — APPROVED** [review]
Tested on TS3. Re-approve.

**Pull request merged**